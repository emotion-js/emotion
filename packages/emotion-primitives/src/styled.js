// @flow
import * as React from 'react'
import { View, Text, Image, StyleSheet } from 'react-primitives'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign
} from './utils'
import { getStyles } from './getStyles'
import { convertToRNStyles } from './convertToRNStyles'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'

function isPrimitiveComponent(component: React.ElementType) {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return true
    }
  }
  return false
}

// Evaluate the styles and convert them to React Native using styled component context
function evalStyles(context, Comp, styles) {
  // Convert styles with or without interpolations to React Native (StyleSheet.create)
  let otherStyles = convertToRNStyles.call(context, styles)

  return getStyles.call(context, otherStyles, context.props)
}

type State = {
  theme: Object
} | null

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */
export function createStyled(component: React.ElementType) {
  let isPrimitive = isPrimitiveComponent(component)
  let pick = isPrimitive
    ? testPickPropsOnPrimitiveComponent
    : testPickPropsOnOtherComponent
  return function createStyledComponent(...styles: *) {
    class Styled extends React.Component<*, State> {
      unsubscribe: number | void
      mergedProps: Object

      static withComponent = (newComponent: React.ElementType) =>
        createStyled(newComponent)(...styles)

      componentWillMount() {
        if (this.context[channel] !== undefined) {
          this.unsubscribe = this.context[channel].subscribe(
            setTheme.bind(this)
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe !== undefined) {
          this.context[channel].unsubscribe(this.unsubscribe)
        }
      }

      render() {
        const { props, state } = this

        // Similar to create-emotion-styled component implementation
        this.mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: (state !== null && state.theme) || props.theme || {}
        })

        const emotionStyles = evalStyles(this, Styled, styles)

        return React.createElement(
          component,
          pickAssign(pick, {}, props, {
            ref: props.innerRef,
            style: StyleSheet.flatten(emotionStyles)
          })
        )
      }
    }

    Styled.contextTypes = contextTypes

    Styled.displayName = `emotion(${getDisplayName(component)})`

    return Styled
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
