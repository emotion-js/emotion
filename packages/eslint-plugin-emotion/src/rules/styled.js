module.exports = {
  meta: {
    fixable: 'code'
  },
  create(context) {
    return {
      ImportDeclaration(node) {
        if (
          node.source.value === 'react-emotion' ||
          node.source.value === 'preact-emotion'
        ) {
          let newImportPath =
            node.source.value === 'react-emotion'
              ? '@emotion/styled'
              : '@emotion/preact-styled'
          let quote = node.source.raw.charAt(0)
          context.report({
            node: node.source,
            message: `styled should be imported from ${newImportPath}`,
            fix:
              node.specifiers.length === 1 &&
              node.specifiers[0].type === 'ImportDefaultSpecifier'
                ? fixer => {
                    return fixer.replaceText(
                      node.source,
                      quote + newImportPath + quote
                    )
                  }
                : undefined
          })
        }
      }
    }
  }
}
