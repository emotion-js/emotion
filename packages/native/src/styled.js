import { StyleSheet } from 'react-native'
import { createStyled } from '@emotion/primitives-core'

/**
 * a function that returns a styled component which render styles in React Native
 */
let styled = createStyled(StyleSheet)

export { styled }
