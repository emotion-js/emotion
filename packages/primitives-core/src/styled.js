// @flow
import * as React from 'react'
import { testAlwaysTrue, pickAssign, interleave } from './utils'
import { ThemeContext } from '@emotion/core'
import { createCss } from './css'

let defaultPickTest = prop => prop !== 'theme' && prop !== 'innerRef'

type options = {
  getShouldForwardProp: (cmp: React.ElementType) => (prop: string) => boolean
}

export function createStyled(
  StyleSheet: Object,
  { getShouldForwardProp = () => defaultPickTest }: options = {}
) {
  const css = createCss(StyleSheet)

  return function createEmotion(component: React.ElementType) {
    let pickTest = getShouldForwardProp(component)

    return function createStyledComponent(...rawStyles: *) {
      let styles

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles)
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      // $FlowFixMe
      let Styled = React.forwardRef((props, ref) => {
        return (
          <ThemeContext.Consumer>
            {theme => {
              let mergedProps = pickAssign(testAlwaysTrue, {}, props, {
                theme: props.theme || theme
              })
              let stylesWithStyleProp = styles
              if (props.style) {
                stylesWithStyleProp = styles.concat(props.style)
              }
              const emotionStyles = css.apply(mergedProps, stylesWithStyleProp)

              if (process.env.NODE_ENV !== 'production' && props.innerRef) {
                console.error(
                  'innerRef is no longer supported, please use ref instead'
                )
              }

              return React.createElement(
                component,
                pickAssign(pickTest, {}, props, {
                  ref: ref,
                  style: emotionStyles
                })
              )
            }}
          </ThemeContext.Consumer>
        )
      })
      // $FlowFixMe
      Styled.withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      Styled.displayName = `emotion(${getDisplayName(component)})`

      return Styled
    }
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
