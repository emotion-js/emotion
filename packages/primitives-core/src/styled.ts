import * as React from 'react'
import { interleave } from './utils'
import { ThemeContext } from '@emotion/react'
import { createCss } from './css'
import { AbstractStyleSheet } from './types'

let testOmitPropsOnComponent = (prop: string) =>
  prop !== 'theme' && prop !== 'as'

interface CreateStyledOptions {
  getShouldForwardProp(cmp: React.ElementType): (prop: string) => boolean
}

interface StyledOptions {
  shouldForwardProp?(prop: string): boolean
}

type StyledProps = Record<string, unknown> & {
  as?: React.ElementType
}

export function createStyled(
  StyleSheet: AbstractStyleSheet,
  options?: CreateStyledOptions
) {
  const getShouldForwardProp =
    options?.getShouldForwardProp ?? (() => testOmitPropsOnComponent)

  const css = createCss(StyleSheet)

  return function createEmotion(
    component: React.ElementType,
    options?: StyledOptions
  ) {
    let shouldForwardProp =
      options && options.shouldForwardProp
        ? options.shouldForwardProp
        : undefined
    let defaultShouldForwardProp =
      shouldForwardProp || getShouldForwardProp(component)
    let shouldUseAs = !defaultShouldForwardProp('as')

    return function createStyledComponent(...rawStyles: any[]) {
      let styles: any[]

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles as [any, ...any[]])
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      let Styled = React.forwardRef<unknown, StyledProps>((props, ref) => {
        const finalTag = (shouldUseAs && props.as) || component

        let mergedProps = props
        if (props.theme == null) {
          mergedProps = {}
          for (let key in props) {
            mergedProps[key] = props[key]
          }
          mergedProps.theme = React.useContext(ThemeContext)
        }

        let finalShouldForwardProp =
          shouldUseAs && shouldForwardProp === undefined
            ? getShouldForwardProp(finalTag)
            : defaultShouldForwardProp

        let newProps: Record<string, unknown> = {}

        for (let key in props) {
          if (shouldUseAs && key === 'as') continue

          if (finalShouldForwardProp(key)) {
            newProps[key] = props[key]
          }
        }

        newProps.style = [css.apply(mergedProps, styles), props.style]
        newProps.ref = ref

        return React.createElement(finalTag, newProps)
      })

      Styled.displayName = `emotion(${getDisplayName(component)})`

      const withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      const castedStyled = Styled as typeof Styled & {
        withComponent: typeof withComponent
      }

      castedStyled.withComponent = withComponent

      return castedStyled
    }
  }
}

const getDisplayName = (
  primitive: string | { displayName?: string; name?: string }
) =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
