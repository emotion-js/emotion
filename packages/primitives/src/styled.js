// @flow
import * as React from 'react'
import { View, Text, Image } from 'react-primitives'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign,
  interleave
} from './utils'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'
import { css } from './css'

function isPrimitiveComponent(component: React.ElementType) {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return true
    }
  }
  return false
}

type State = {
  theme: Object
} | null

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */

type CreateStyledComponent = (...styles: any) => React.ElementType

type BaseCreateStyled = (tag: React.ElementType) => CreateStyledComponent

export type CreateStyled = BaseCreateStyled & {
  View: CreateStyledComponent,
  Text: CreateStyledComponent,
  Image: CreateStyledComponent
}

// $FlowFixMe
let createStyled: CreateStyled = function createStyled(
  component: React.ElementType
) {
  let isPrimitive = isPrimitiveComponent(component)
  let pickTest = isPrimitive
    ? testPickPropsOnPrimitiveComponent
    : testPickPropsOnOtherComponent
  return function createStyledComponent(...rawStyles: *) {
    let styles

    if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
      styles = rawStyles
    } else {
      styles = interleave(rawStyles)
    }
    class Styled extends React.Component<*, State> {
      unsubscribe: number | void
      mergedProps: Object

      static withComponent = (newComponent: React.ElementType) =>
        createStyled(newComponent)(...styles)

      componentWillMount() {
        if (this.context[channel] !== undefined) {
          this.unsubscribe = this.context[channel].subscribe(
            setTheme.bind(this)
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe !== undefined) {
          this.context[channel].unsubscribe(this.unsubscribe)
        }
      }

      render() {
        const { props, state } = this

        // Similar to create-emotion-styled component implementation
        this.mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: (state !== null && state.theme) || props.theme || {}
        })
        let stylesWithStyleProp = styles
        if (props.style) {
          stylesWithStyleProp = styles.concat(props.style)
        }
        const emotionStyles = css.apply(this, stylesWithStyleProp)

        return React.createElement(
          component,
          pickAssign(pickTest, {}, props, {
            ref: props.innerRef,
            style: emotionStyles
          })
        )
      }
    }

    Styled.contextTypes = contextTypes

    Styled.displayName = `emotion(${getDisplayName(component)})`

    return Styled
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'

export { createStyled }
