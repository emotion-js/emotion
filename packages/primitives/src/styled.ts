import * as React from 'react'
import { StyleSheet, View, Text, Image } from 'react-primitives'
import { createStyled } from '@emotion/primitives-core'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'

function getShouldForwardProp(
  component: React.ElementType
): (prop: string) => boolean {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return testPickPropsOnPrimitiveComponent
    }
  }
  return testPickPropsOnOtherComponent
}

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */
type CreateStyledComponent = (...styles: any) => React.FC<any> & {
  withComponent: (component: any) => React.FC<any>
}

/** @internal */
export type BaseStyled = (tag: React.ElementType) => CreateStyledComponent

export type Styled = BaseStyled & {
  View: CreateStyledComponent
  Text: CreateStyledComponent
  Image: CreateStyledComponent
}

let styled: BaseStyled = createStyled(StyleSheet, { getShouldForwardProp })

export { styled }
