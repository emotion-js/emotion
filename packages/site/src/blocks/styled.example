const Image = ({ className, src }) => <img src={src} className={className} />

const Avatar = styled(Image)`
  width: 96px;
  height: 96px;
  border-radius: 50%;
  transition: transform 400ms ease-in-out;

  &:hover {
    transform: scale(1.2);
  }
`

render(<Avatar src={logoUrl} />)
