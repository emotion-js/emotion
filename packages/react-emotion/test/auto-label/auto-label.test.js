import React from 'react'
import renderer from 'react-test-renderer'
import styled, { css, flush, sheet } from 'react-emotion'

import { hiDPI, lighten, modularScale } from 'polished'

describe('styled with autoLabel', () => {
  beforeEach(() => flush())
  test('composition', () => {
    const fontSize = '20px'

    const cssA = css`
      color: blue;
    `

    const cssB = css`
      ${cssA};
      color: red;
    `

    const BlueH1 = styled('h1')`
      ${cssB};
      color: blue;
      font-size: ${fontSize};
    `

    const FinalH2 = styled(BlueH1)`
      font-size: 32px;
    `

    const tree = renderer
      .create(
        <FinalH2 scale={2} className={'legacy__class'}>
          hello world
        </FinalH2>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('composition with objects', () => {
    const cssA = {
      color: lighten(0.2, '#000'),
      fontSize: modularScale(1),
      [hiDPI(1.5)
        .replace('\n', ' ')
        .trim()]: { fontSize: modularScale(1.25) },
    }

    const cssB = css`
      ${cssA};
      height: 64px;
    `

    const H1 = styled('h1')`
      ${cssB};
      font-size: ${modularScale(4)};
    `

    const H2 = styled(H1)`
      font-size: 32px;
    `

    const tree = renderer
      .create(
        <H2 scale={2} className={'legacy__class'}>
          hello world
        </H2>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('higher order component', () => {
    const fontSize = 20
    const Content = styled('div')`
      font-size: ${fontSize}px;
    `

    const squirtleBlueBackground = css`
      background-color: #7fc8d6;
    `

    const flexColumn = Component => {
      const NewComponent = styled(Component)`
        ${squirtleBlueBackground};
        background-color: '#343a40';
        flex-direction: column;
      `

      return NewComponent
    }

    const ColumnContent = flexColumn(Content)

    const tree = renderer.create(<ColumnContent />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })
  test('displayName', () => {
    const SomeComponent = styled.div`
      color: hotpink;
    `
    expect(SomeComponent.displayName).toBe('SomeComponent')
  })
})
