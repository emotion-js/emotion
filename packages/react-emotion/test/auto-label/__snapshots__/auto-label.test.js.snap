// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled with autoLabel composition 1`] = `
.glamor-0 {
  color: blue;
  color: red;
  color: blue;
  font-size: 20px;
  font-size: 32px;
}

<h1
  className="legacy__class glamor-0"
  scale={2}
>
  hello world
</h1>
`;

exports[`styled with autoLabel composition 2`] = `
".css-1nvqkkc-cssA {
  color: blue;
}

.css-1xoijq0-cssB {
  color: blue;
  color: red;
}

.css-1ncw0w-BlueH1-FinalH2 {
  color: blue;
  color: red;
  color: blue;
  font-size: 20px;
  font-size: 32px;
}"
`;

exports[`styled with autoLabel composition with objects 1`] = `
.glamor-0 {
  color: #333;
  font-size: 1.333em;
  height: 64px;
  font-size: 3.157334518321em;
  font-size: 32px;
}

@media only screen and (-webkit-min-device-pixel-ratio:1.5), only screen and (min--moz-device-pixel-ratio:1.5), only screen and (-o-min-device-pixel-ratio:1.5/1), only screen and (min-resolution:144dpi), only screen and (min-resolution:1.5dppx) {
  .glamor-0 {
    font-size: 1.4323121856191332em;
  }
}

<h1
  className="legacy__class glamor-0"
  scale={2}
>
  hello world
</h1>
`;

exports[`styled with autoLabel composition with objects 2`] = `
".css-10adwdu-cssB {
  color: #333;
  font-size: 1.333em;
  height: 64px;
}

@media only screen and (-webkit-min-device-pixel-ratio:1.5), only screen and (min--moz-device-pixel-ratio:1.5), only screen and (-o-min-device-pixel-ratio:1.5/1), only screen and (min-resolution:144dpi), only screen and (min-resolution:1.5dppx) {
  .css-10adwdu-cssB {
    font-size: 1.4323121856191332em;
  }
}

.css-13djram-H1-H2 {
  color: #333;
  font-size: 1.333em;
  height: 64px;
  font-size: 3.157334518321em;
  font-size: 32px;
}

@media only screen and (-webkit-min-device-pixel-ratio:1.5), only screen and (min--moz-device-pixel-ratio:1.5), only screen and (-o-min-device-pixel-ratio:1.5/1), only screen and (min-resolution:144dpi), only screen and (min-resolution:1.5dppx) {
  .css-13djram-H1-H2 {
    font-size: 1.4323121856191332em;
  }
}"
`;

exports[`styled with autoLabel higher order component 1`] = `
.glamor-0 {
  font-size: 20px;
  background-color: #7fc8d6;
  background-color: '#343a40';
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="glamor-0"
/>
`;

exports[`styled with autoLabel higher order component 2`] = `
".css-l19ikj-squirtleBlueBackground {
  background-color: #7fc8d6;
}

.css-t14qr0-Content-NewComponent {
  font-size: 20px;
  background-color: #7fc8d6;
  background-color: '#343a40';
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}"
`;
